### ISSUE TITLE
Layers/Extends Support Tracker

### ISSUE THREAD

**Issue by pi0** [nuxt/nuxt#13367]
Nuxt layers are a powerful feature that you can use to share and reuse partial Nuxt applications within a monorepo, or from a git repository or npm package.

Docs:
- Usage: https://nuxt.com/docs/getting-started/layers
- Authoring: https://nuxt.com/docs/guide/going-further/layers

This issue is used to track the progress of enhancements and known issues in the roadmap.

Features:
- [x] Initial support (#3008)
- [x] `components/` dir (#3108)
- [x] `composables/` dir (https://github.com/nuxt/framework/pull/3423)
- [x] `plugins/` dir (https://github.com/nuxt/framework/pull/3462)
- [x] `server/` dir (#3673) [need refactor for nitropack]
- [x] `pages/` dir (#3783)
- [x] `middleware/` dir (#3783)
- [x] `layouts/` dir (through https://github.com/nuxt/framework/pull/3610)
- [x] `public/` dir (nitropack pending but could be workaround to pick one layer)
- [x] `app/` dir (`app/router.options`) (https://github.com/nuxt/framework/pull/3939)
- [x] app.vue (https://github.com/nuxt/framework/pull/6228)
- [x] hooks for extending dirs
- [x] Usage and authoring docs nuxt/framework#9405
- [x] Merge `hook` config key: https://github.com/nuxt/nuxt/pull/24639
- [x] install cloned dependencies (https://github.com/unjs/c12/issues/51)
- [ ] #30941
- [ ] #21635

Bugs:
- [x] Auto imports for nitro layers (#9384)

**Comment by nonlinearcom** [comment]
Does anyone know a workaround to make extends work with imported modules? or is this not available at the moment?
https://github.com/nuxt/nuxt/issues/18448

**Comment by szulcus** [comment]
Hi, your work is awesome! ❤️
Can I use `vite.css.preprocessorOptions.scss.additionalData` option in nuxt layer?

**Comment by adam-code-labx** [comment]
> Does anyone know a workaround to make extends work with imported modules? or is this not available at the moment? #18448

I'm facing the same problem here at the moment. The only workaround is to make sure that all layer dependencies have been correctly installed in the parent core application. It's certainly frustrating, especially when we're trying to keep the theme dependencies separate from the core app ones.

Are dynamic images a factor in this issue? Currently, images have to be manually transferred to the core application, as they can not be resolved in the theme layer.

For example `~/assets/` path in our composable does not resolve to the theme layer.

``` js
export default function useAsset(path: string): string {
  // ignore presets directory when using meta glob
  const assets = import.meta.glob(['~/assets/**/*', '!**/presets'], {
    eager: true,
    import: 'default',
  })
  // @ts-expect-error: wrong type info
  return assets[`/assets/${path}`]
}
```

**Comment by xibman** [comment]
@adam-code-labx  

I see something on this in this thread. 
You need to use this:

```ts
const { resolve } = createResolver(import.meta.url);

export default function useAsset(path: string): string {
  // ignore presets directory when using meta glob
  const assets = import.meta.glob([resolve('~/assets/**/*'), '!**/presets'], {
    eager: true,
    import: 'default',
  })
  // @ts-expect-error: wrong type info
  return assets[`/assets/${path}`]
}
```

**Comment by pbtkhoa** [comment]
the layers feature only used for the override; I mean, if we create a file with the same location in the main project, it will override, so Do we have any solution to extend (reuse all the code and functions)?

EG:
I have a file called `pages/account.vue`  in `@package/test`
```
<template>
    <div>
        Account 1
        <slot name="content" />
    </div>
</template>
<script setup>
const parentMethod = () => {
    console.log('parent')
}
...many methods
</script>
```

And In my main project, I also have a file `pages/account.vue`; how can I override only `parentMethod` method in this component?

**Comment by adam-code-labx** [comment]
Hi there @pbtkhoa
I experienced the same issue and was able to get around it by adding the hook to our theme layer component nuxt config. 
I believe this will currently only handle first-level pages and not nested.

``` js
   hooks: {
    'pages:extend': (pages) => {
      const files = fs.readdirSync(resolve(__dirname, 'pages'))
      files.forEach((file) => {
        const name = file.replace('.vue', '')
        pages.push({
          name,
          path: `/${name === 'index' ? '' : name}`,
          file: resolve(__dirname, `pages/${file}`),
        })
      })
    },
  },
```

**Comment by warflash** [comment]
👋
just curious what exactly `Multi-app support` does relate to.
I'm trying to deploy multiple nuxt projects on the same domain with e.g. `/admin` routing to a standalone admin app and `/store` routing to a store app. 
Both apps extend off of a base project for shared components which works well, just wondering how one would go about the "prefix based app routing" and whether this multi app support checklist entry could be the solution if one doesnt exist at the moment 
Thank you!

**Comment by tcm151** [comment]
> What I suggest is to have a similar way to configure your extends as components does:
> 
> ```ts
> defineNuxtConfig({
>    extends: [
>       './base', // Would be fully extended
>       {
>             path: './ui',
>             middleware: false, // Would skip middleware extending
>             pages: false, // Would skip pages extending
>             // All of the rest are set to `true` by default
>         }
>    ]
> })
> ```

I've read thru this thread and I don't seem to see a definitive answer. I am very interested in being able to ignore certain directories/files when extending from a separate project. For example I have my server/API in it's own project and have it so that certain pages/routes do not get imported when extending from that project, in say my UI. 

I thought that maybe including those files in the `.nuxtignore` might do the trick, but that doesn't seem to work. Then I stumbled across this thread. 

I am also interested in being able to include custom directories that are specified in the `nuxt.config.ts`. For example `imports: { dirs: ["services/**"] }` is specified in one project, then it would also be imported when extending from that project.

**Comment by Atinux** [comment]
@tcm151 it's harder to support options like this, but what you can do is splitting your `./base` into different layers:
- `base` that can only have components
- `base-pages` 
- `base-middleware`
Etc.

The thing is that any modules could leverage extends for it's own system, example: `@nuxtjs/tailwindcss` uses it for `tailwind.config` in multiple layers to extend.

I think that for the `imports` key it could be supported with a pull request :)

**Comment by ennioVisco** [comment]
Sharing assets is still a major limiting factor to authoring themes as layers. 
It seems though there is no PR on that (or the more general `Named layer alias`.
Is it problematic in some way?

**Comment by szulcus** [comment]
@ennioVisco I'm waiting for that too

**Comment by Atinux** [comment]
When you mean assets you are talking about the `public/` directory?

**Comment by ennioVisco** [comment]
> When you mean assets you are talking about the `public/` directory?

Well no, I'm talking about the `assets` dir. 
Ideally theme's css and images should be accessible from extenders

**Comment by ennioVisco** [comment]
More specifically, in my use case, we'd like to have a `ui` layer, but together with components and composables, it should ideally also export some styling configs, e.g. a set of `sass` mixins

**Comment by MikaKoskiola** [comment]
Any plans to support extending content?

**Comment by manniL** [comment]
@MikaKoskiola AFAIK this should already work - https://github.com/nuxt/content/pull/1404/files

If not, raise an issue in the content module repo.

**Comment by HapLifeMan** [comment]
Hello, glad to see it's moving forward here!
Any information if we can already have it working like this? https://github.com/nuxt/rfcs/issues/30#issuecomment-506716083

**Comment by Atinux** [comment]
From your comment @HapLifeMan 

You can conditionally add a layer based on an env variable, it could do the trick.

**Comment by oemer-sellwerk** [comment]
> 👋 just curious what exactly `Multi-app support` does relate to. I'm trying to deploy multiple nuxt projects on the same domain with e.g. `/admin` routing to a standalone admin app and `/store` routing to a store app. Both apps extend off of a base project for shared components which works well, just wondering how one would go about the "prefix based app routing" and whether this multi app support checklist entry could be the solution if one doesnt exist at the moment Thank you!

I am also interested in this. What does `Multi-app support` mean?

What I would really like to do is to create layer system where teams can work independently and domain-driven, like this:
`base` - Includes all shared files, components, modules etc.
`feature-1` - Includes pages for feature-1, where only Team 1 works with (extends base)
`feature-2` - Includes pages for feature-2, where only Team 2 works with (extends base)
`main` - Includes Auth pages, landing pages and feature pages (extends base, feature-1, feature-2)

Is that what's meant with `Multi-App`? Does this architecture make sense?

**Comment by AndrewBogdanovTSS** [comment]
I wonder if there is a possibility to extend base layer not sequentially but in parallel like it was described in the above comment, so feature-1 and feature-2 could extend from base but not from one another but at the same time any domain that goes after features would extend all domains, so imagine such a structure:
base -> feature-1| feature-2 -> theme-1 | theme-2 | theme-3
Where features and theme are parallel to each other

**Comment by Atinux** [comment]
I believe multi apps lives above the layers so they can have their own layers.

**Comment by pbtkhoa** [comment]
I'm asking myself how can we extend and reuse all methods/views:

`feature-1` - > I have component A for example: 
```vue
<template>
    <div>
        Account 1
        <slot-to-let-children-modify>
    </div>
</template>
<script setup>
const parentMethod = () => {
    console.log('parent')
}
</script>
```

`main` - > extends `feature-1` but wants to modify component A (for example change view, methods, computed...) 
```vue
<template>
   <slot-to-let-children-modify>
      Add to parent
   </slot-to-let-children-modify>
</template>
<script setup>
const childMethod = () => {
    // ... I want to call parentMethod here
}
</script>
```

So, my question is, how is it possible with the `Layer` feature?

**Comment by Atinux** [comment]
The component in `main` layer will overwrite the component in `feature-1` layer, it does not extend it.

If you want to do so, just name it differently in `feature-1` (ex: `ComponentAExtended`) and extend the component A from it.

**Comment by baixiaoyu2997** [comment]
> 

Or can we provide an option that can be ignored in both development and production environments? Like `ignore`, but it only takes effect at build time.

**Comment by baixiaoyu2997** [comment]
> @tcm151 it's harder to support options like this, but what you can do is splitting your `./base` into different layers:
> 
> * `base` that can only have components
> * `base-pages`
> * `base-middleware`
>   Etc.
> 
> The thing is that any modules could leverage extends for it's own system, example: `@nuxtjs/tailwindcss` uses it for `tailwind.config` in multiple layers to extend.
> 
> I think that for the `imports` key it could be supported with a pull request :)

Or can we provide an option that can be ignored in both development and production environments? Like ignore, but it only takes effect at build time.

**Comment by adam-code-labx** [comment]
I am currently facing a similar issue with the components of my app that I haven't been able to solve.

```js
const component = defineAsyncComponent(
  () => import(`@/components/widgets/${props.widget.type}/${props.widget.component}.vue`),
) as any
```
Specifically, I have an app that serves as a theme, and I have been successful in dynamically loading its components using defineAsyncComponent. However, I have encountered a new challenge where I need to relocate all the builder components and core components to a separate package or 'layer' so I can reuse for multiple theme projects.

Unfortunately, I have not found a way to access the theme components in this new layer. Moreover, the '@' symbol is not working from the new layer, which has made it difficult for me to load the components. To work around this issue, I have resorted to using relative paths.

In this particular scenario, I need to check whether a component has been overridden from our main theme app. If it hasn't, I would like to load the default component from our builder. However, due to the current limitations, I haven't been able to achieve this goal.

If anyone has any suggestions or guidance on how to approach this issue, I would greatly appreciate it. Thank you

**Comment by manniL** [comment]
> To work around this issue, I have resorted to using relative paths.

Indeed, that is a known restriction for now and also documented.
If the component is in the same place as before, it should override the previous layer's component automatically.

**Comment by genu** [comment]
> Did you anyone try to use layers with UnoCSS? it looks I am not able to replace/extend f.e. shortcuts.
> 
> 
> 
> Also when I use separate config file `unocss.config.ts` instead of inline config in `nuxt.config.ts` it doesn't load it overall.
> 
> When I use standalone config, it ignores all previous layers configs (whether it is inline or standalone) so when use unocss.config.ts on the last layer, everything else is ignored. 

How are you loading the Unocss config file?

**Comment by KeJunMao** [comment]
> Did you anyone try to use layers with UnoCSS? it looks I am not able to replace/extend f.e. shortcuts.
> 
> Also when I use separate config file `unocss.config.ts` instead of inline config in `nuxt.config.ts` it doesn't load it overall. When I use standalone config, it ignores all previous layers configs (whether it is inline or standalone) so when use unocss.config.ts on the last layer, everything else is ignored.

try this

```ts
// your layer nuxt.config.ts
import { createResolver } from "@nuxt/kit";
const { resolve } = createResolver(import.meta.url);

export default defineNuxtConfig({
  modules: [
    "@unocss/nuxt",
  ],
  unocss: {
    configFile: resolve("./unocss.config.ts"),
  },
})
```

**Comment by kleinpetr** [comment]
> How are you loading the Unocss config file?

I tried inline config in `nuxt.config.ts` and also unocss.config.ts, when there should not be necessary to import them in nuxt config (it works only for the last layer), I've figured out, that if you use unocss config inside the `nuxt.config.ts` it works a bit, it merge all shortcuts from all layers, BUT it cannot replace shortcuts, so it means when you have `baseLayer` with shortcut `btn: bg-blue` and then you rewrite this shortcut on the `secondLayer` with `btn: bg-red` it still shows blue

**Comment by chrisfoss1** [comment]
> I am currently facing a similar issue with the components of my app that I haven't been able to solve.
> 
> ```js
> const component = defineAsyncComponent(
>   () => import(`@/components/widgets/${props.widget.type}/${props.widget.component}.vue`),
> ) as any
> ```
> 
> Specifically, I have an app that serves as a theme, and I have been successful in dynamically loading its components using defineAsyncComponent. However, I have encountered a new challenge where I need to relocate all the builder components and core components to a separate package or 'layer' so I can reuse for multiple theme projects.
> 
> Unfortunately, I have not found a way to access the theme components in this new layer. Moreover, the '@' symbol is not working from the new layer, which has made it difficult for me to load the components. To work around this issue, I have resorted to using relative paths.
> 
> In this particular scenario, I need to check whether a component has been overridden from our main theme app. If it hasn't, I would like to load the default component from our builder. However, due to the current limitations, I haven't been able to achieve this goal.
> 
> If anyone has any suggestions or guidance on how to approach this issue, I would greatly appreciate it. Thank you

Regarding your issue with relative paths, I worked around that by having each layer define its own alias instead of using "@".

```
import { createResolver } from "@nuxt/kit";
const { resolve } = createResolver(import.meta.url);

import { type NuxtConfig, defineNuxtConfig } from "nuxt/config";

// https://v3.nuxtjs.org/api/configuration/nuxt.config
const config: NuxtConfig = defineNuxtConfig({
  alias: { "#component-library": resolve("./src/") },
  ...etc
});
```
Then instead of referencing `@/path/to/layer/file` I use `#component-libary/path/to/layer/file`.  Your tsconfig needs to extend `.nuxt/tsconfig.json`, since that is where the alias will get set up for typescript.

**Comment by kleinpetr** [comment]
Also does anyone facing VsCode auto import issues when using monorepo with pnpm? Build and preview works well. So its mainly about vsCode setup I guess. Thanks

![image](https://user-images.githubusercontent.com/17269309/227734219-f5c51e9e-16c1-4c55-adb2-69c1b8f54f1b.png)

**Comment by manniL** [comment]
@kleinpetr Run `nuxi preprare` in each layer folder ☺️

**Comment by kleinpetr** [comment]
it runs `"postinstall": "nuxt prepare .playground",` for each layer when I run root `pnpm i` for example

![image](https://user-images.githubusercontent.com/17269309/227735395-f2e4cbc0-b0b7-4f9c-ae4c-faa5f329ed52.png)

**Comment by manniL** [comment]
How does your tsconfig look like?

**Comment by kleinpetr** [comment]
tsconfig in each layer
{
  "extends": "./.playground/.nuxt/tsconfig.json"
}
```

**Comment by adam-code-labx** [comment]
Thank you, @chrisfoss1 for sharing your solution on defining aliases within the same layer. 
This is certainly a very useful technique for dynamically declaring aliases which I never even considered 💪

However in my case I needed to access the core app from the components layer of my theme. 
After some experimentation. I discovered that the core app can be accessed via the `/_nuxt/` directory.

I have now created a composable solution that allows me to use the components layer as the default components, while also providing the ability to override them from my main app.

If anyone else is facing a similar issue. I hope this solution will be helpful for you.

```ts
import { Component, defineAsyncComponent } from 'vue'
type ComponentLoader = () => Promise<{ default: Component }>

export default function useComponent(componentPath: string, rootComponentDir = '/_nuxt/components/'): ComponentLoader {
  const fallback = defineAsyncComponent(() => import(`../../components/${componentPath}`)) as ComponentLoader
  return defineAsyncComponent({
    loader: async () => {
      try {
        return await import(`${rootComponentDir}${componentPath}`)
      } catch (error) {
        return fallback
      }
    },
  }) as ComponentLoader
}
```

Example vue component usage:

```vue
<script setup lang="ts">
import useComponent from '../../composables/useComponent'
const props = defineProps<{
  widget: Record<string, any>
}>()

const loadedComponent = useComponent(`widgets/${props.widget.type}/${props.widget.component}.vue`)
</script>

<template>
  <div>
    <Component :is="loadedComponent" :widget="widget">
      <slot />
    </Component>
  </div>
</template>
```

**Comment by kleinpetr** [comment]
@manniL but its kind a weird, it seems that vsCode cannot resolve the relative path :thinking: 

![image](https://user-images.githubusercontent.com/17269309/227738670-97ac17d3-31c1-4304-861f-a4f54d449a21.png)

**Comment by manniL** [comment]
@kleinpetr Then you have to run `nuxi prepare` in the `.playground` folder so the `.nuxt` folder will created there and the TSConfig has the reference :)

**Comment by kleinpetr** [comment]
> @kleinpetr Then you have to run `nuxi prepare` in the `.playground` folder so the `.nuxt` folder will created there and the TSConfig has the reference :) 

But the tsconfig.json inside .playground is generated well, it contains all. So that's the issue probably only about vscode setting 🤔

**Comment by kleinpetr** [comment]
@manniL as you can see when I run dev server or even when I run `nuxi prepare` it loads correct for a while but then show the error again.

![monorepo-import](https://user-images.githubusercontent.com/17269309/227916981-120d08cd-f958-41f9-8e56-5016c0b2b601.gif)

**Comment by kleinpetr** [comment]
> > Did you anyone try to use layers with UnoCSS? it looks I am not able to replace/extend f.e. shortcuts.
> > Also when I use separate config file `unocss.config.ts` instead of inline config in `nuxt.config.ts` it doesn't load it overall. When I use standalone config, it ignores all previous layers configs (whether it is inline or standalone) so when use unocss.config.ts on the last layer, everything else is ignored.
> 
> try this
> 
> ```ts
> // your layer nuxt.config.ts
> import { createResolver } from "@nuxt/kit";
> const { resolve } = createResolver(import.meta.url);
> 
> export default defineNuxtConfig({
>   modules: [
>     "@unocss/nuxt",
>   ],
>   unocss: {
>     configFile: resolve("./unocss.config.ts"),
>   },
> })
> ```
> 
> Edit:
> 
> sorry, I misunderstood it. This is an example of how to load an external file

Thanks! with the `resolve` it loads the uno config in layers which extends from, and also I am able to rewrite `unocss.config.ts` in next layers. 

I am getting another err within the `import.meta` which is

**Comment by kleinpetr** [comment]
I've pushed a minimal repo with reproduce https://github.com/kleinpetr/nuxt-layers

when I run `pnpm i` in the root, the app `nuxt.config.ts` looks fine, but `nuxt.config.ts` inside `my-layer` does not.

**Comment by manniL** [comment]
@kleinpetr best would be an extract discussion or issue so we can keep this one lean ☺️

**Comment by kleinpetr** [comment]
for sure, I am gonna do it. Thx :pray:

**Comment by kvanska** [comment]
I'm trying to extend /content .md files from my base layer. for some reason it's not working. 

Small repo: [https://codesandbox.io/p/github/kvanska/content-app/master](https://codesandbox.io/p/github/kvanska/content-app/master)

"/docs" and "/docs/about" don't find the content files even when they are added to the base layers nuxt.config.ts:

```
export default defineNuxtConfig({
  imports: {
    dirs: ['content/docs']
  }
})
```

**Comment by robinscholz** [comment]
I am currently running into issues with overriding the default way components are imported within my base layer. I'd like to remove the pathPrefix, but adding any sort of `components` config to the `nuxt.config.ts` of the base-layer results in the top layer app breaking, since it can’t resolve any of the components defined within the base layer it extends from.

Anyone else run into similiar issues?

**Comment by tcm151** [comment]
If you are okay with your different layers sharing ALL dependencies then you can use the `modulesDir` property in the `nuxt.config.ts` to point them all to the same place. Consider the following folder structure and config:

```txt
|- node_modules
|- core
|--- pages
|--- server
|--- etc
|--- nuxt.config.ts
|- project-1
|--- etc
|--- nuxt.config.ts
|- project-2
|--- etc
|--- nuxt.config.ts
|- package.json
|- yarn.lock
```
```js
import { createResolver } from '@nuxt/kit'
const { resolve } = createResolver(import.meta.url);

export default defineNuxtConfig({
    alias: {
        '@core': resolve('./')
    },
    modulesDir: [
        '../node-modules'
    ],
})
```

Using the suggestion of @chrisfoss1 for the layer aliases, then you can very easily access the core project here in this example from either of the two other projects. Using `@core/assets` or `@core/path/to/anything` (you even get directory intellisense in vscode).

As well when extending from the config in the other projects, as long as you use the `@core/` alias then you don't need to do any additional resolves to get the proper path to the object. For example I am importing sass style sheets into every component in the core project, and they are automatically imported the same way in my other two projects.

```js
export default defineNuxtConfig({
    alias: {
        '@core': resolve('./')
    },
    modulesDir: [
        '../node-modules'
    ],
    css: [
        '@core/assets/scss/styles.scss'
    ],
    vite: {
        css: {
            preprocessorOptions: {
                scss: {
                    additionalData: '@use "@core/assets/scss/modules" as *;\n'
                }
            }
        }
    }
})
```

**Comment by kleinpetr** [comment]
Hey I've created an Issue when I post my findings about using UnoCSS within layers. https://github.com/unocss/unocss/issues/2430

**Comment by kleinpetr** [comment]
I am facing an issue with `/public` assets when inheriting over 2+ layers. In the third layer there is an error during build.

The structure is

layer1 --> layer2 --> layer3

in my-layer1 I have `Header` components which shows `<img src="/img/placeholder.png>`

1. when I run layer2 which has no `/public` folder -> it works
2. when I run layer3 which has no `/public` folder -> it doesn't work (error below)

overall dev via Vite works well but build via Rollups does not.

```
[vite]: Rollup failed to resolve import "/img/placeholder.png" from "/home/..../nuxt-layers/layers/layer1/components/Header.vue?vue&type=script&setup=true&lang.ts".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
```

**Comment by enkot** [comment]
@pi0 It would be great to add possibility to extend `server/assets` folder in layers.

**Comment by pi0** [comment]
@enkot Do you mind to open an issue in nitro repo to track?

**Comment by ananthachetan** [comment]
> I've pushed a minimal repo with reproduce https://github.com/kleinpetr/nuxt-layers
> 
> when I run `pnpm i` in the root, the `/app/nuxt.config.ts` looks fine, but `nuxt.config.ts` inside `/layers/my-layer` does not.

I also encountered this issue on Webstorm when using layers. Not sure if it is the right fix. I changed the nuxt.config.ts inside `/layers/my-layer` back to 
```
{
    "extends": "./.nuxt/tsconfig.json"
}
```
Modified the package.json of the `/layers/my-layer` to 
```
"postinstall": "nuxi prepare && nuxi prepare .playground"
```
And then did the following in `/layers/my-layer`

nuxi cleanup
nuxi prepare

After this all the ts errors in nuxt.config.ts were resolved.

**Comment by kleinpetr** [comment]
@ananthachetan you can see particular issue with fix here https://github.com/nuxt/nuxt/issues/20155#issuecomment-1501219431

**Comment by akasection** [comment]
I am currently stuck with problem declaring middleware and layout for page layer, which the middleware comes from the Nuxt app rather than the layer. I can't use `definePageMeta` and `appConfig.meta.middleware`  altogether since `definePageMeta` can't read `appConfig` (since it is just a macro).

To illustrate, let's say the layer provides a feature page, and the Nuxt app provides the auth/login middleware and the dashboard layout.

Any way to do so?

**Comment by kleinpetr** [comment]
> I am facing an issue with `/public` assets when inheriting over 2+ layers. In the third layer there is an error during build.
> 
> The structure is
> 
> layer1 --> layer2 --> layer3
> 
> in the layer1 I have `Header` components which shows `<img src="/img/placeholder.png">`
> 
> 1. when I run layer2 which has no `/public` folder -> it works
> 2. when I run layer3 which has no `/public` folder -> it doesn't work (error below)
> 
> overall dev via Vite works well but build via Rollups does not.
> 
> ```
> [vite]: Rollup failed to resolve import "/img/placeholder.png" from "/home/..../nuxt-layers/layers/layer1/components/Header.vue?vue&type=script&setup=true&lang.ts".
> This is most likely unintended because it can break your application at runtime.
> If you do want to externalize this module explicitly add it to
> `build.rollupOptions.external`
> ```

no one facing this issue?

**Comment by danielroe** [comment]
If you are facing a _bug_ the best thing is to open a new issue with a reproduction. This tracker issue is not for reporting bugs but giving an overview of the implementation.

**Comment by luyentm** [comment]
@danielroe can we change order of layers in runtime?

**Comment by CavalcanteLeo** [comment]
> ```ts
> 
> import { createResolver } from "@nuxt/kit";
> const { resolve } = createResolver(import.meta.url);
> ```

I'm

**Comment by vuejstime** [comment]
i am using `gsap` in main project and also in `layer` but when i run `npx nuxi analyze` it will bundle twice, how to fix this?
thanks

**Comment by Dominic-Marcelino** [comment]
I'm currently struggeling with proper TS support in layers. We have some composables that use custom types in layer one. Layer two extends layer one and overrides some of the composables but uses the same types.
 
While the layering-system takes part of most of the logic it would be great to share the types in a better way than overriting or creating a local types file in the same folder. What do you think of adding a `types/ ` folder to the resolved folders?

**Comment by oemer-aran** [comment]
> I'm currently struggeling with proper TS support in layers. We have some composables that use custom types in layer one. Layer two extends layer one and overrides some of the composables but uses the same types.
> 
> While the layering-system takes part of most of the logic it would be great to share the types in a better way than overriting or creating a local types file in the same folder. What do you think of adding a `types/ ` folder to the resolved folders?

I solved most of my problems by following the suggestion on this thread to add an alias to each layer. 
This is how it could help you in your example:
```ts
// /layer-one/nuxt.config.ts
import { createResolver } from "@nuxt/kit"
const { resolve } = createResolver(import.meta.url)

export default defineNuxtConfig({
  alias: { "@layer-one": resolve("./") },
})
```
Now you can import **everything** from Layer One to Layer Two, e.g.:
```ts
// /layer-two/composables/useMyComposable.ts
import  type { ComposableType } from "@layer-one/composables/useMyComposable"

export const userMyComposable(args: ComposableType) { ... }
```

No need to add an extra `/types` folder,  just keep your files co-located and clean.

**Comment by Dominic-Marcelino** [comment]
> I solved most of my problems by following the suggestion on this thread to add an alias to each layer. This is how it could help you in your example:

Thanks, didn't thought about a this!

**Comment by pi0** [comment]
Feel free to open an issue in nitro repository regarding public directory multi layer support. it is expected to work (and there might be one issue already haven't checked)

**Comment by raevilman** [comment]
Given: a `HelloLayer.vue` at location `components/greet/HelloLayer.vue` in the layer app.

The consumer app can _only_ use the component with name `GreetHelloLayer` but not `HelloLayer`.

Is this expected behaviors? 


Similarly, in layer app, 
`components/lvl1/lvl2/lvl3/lvl4/lvl5/Level5Layer.vue` is resulting in `Lvl1Lvl2Lvl3Lvl4Lvl5Level5Layer`

**Comment by danielroe** [comment]
Yes, this is expected (and is also true without layers). If you want the path to be ignored, you can add a `~/components/greet` path to the configured components dirs.

**Comment by Arthur-Chevron** [comment]
I create a blog on another project than my main project but i want to display my articles in my main app and it seems that content directory is not auto import.

Do you know how to auto import content folder from my layer to my app repo ?

**Comment by maxdzin** [comment]
Nice to see the work is continuing for layers support!

@pi0 how about supporting [utils](https://nuxt.com/docs/guide/directory-structure/utils) top-level directory?

**Comment by pi0** [comment]
@maxdzin Feel free to open a PR for utils support from layers! (it should work same as composables)

**Comment by localusercamp** [comment]
What about allowing to extend `extends` through module?

**Comment by pi0** [comment]
> What about allowing to extend extends through module?

Modules can programmatically extend builder API and use nuxt kit utilities to add auto imported dirs such as plugins, server API, etc. We were also internally thinking about introducing another kit utility to make it possible to register a whole auto scanned directory which will be having same structure of a nuxt app / layer. 

But in general, nuxt layers, are more extendable. They are before nuxt build phase and modules initialization starting and provide functionalities modules technically can't.

**Comment by localusercamp** [comment]
> We were also internally thinking about introducing another kit utility to make it possible to register a whole auto scanned directory which will be having same structure of a nuxt app / layer.

Are talking about some kind of `/layers` directory where you can put all your app layers?

**Comment by localusercamp** [comment]
Maybe some kind of pre-build scripts or pre-build modules is the thing...

**Comment by MuhammadM1998** [comment]
Found this after asking on Discord if one can extend configs from a layer and after reading the features list it seems it's not supported.

My use case is I want to have a `base` repo that has core UI components in addition to eslint, prettier configs.
I want to use those UI components in addition to `eslint.config` and `.prettierrc` in `app-1`, `app-2` repos for example to ensure code consistency (I think this is much like a mono repo?).

Adding to @Tahul proposal. Can extending configs be supported? maybe user need to explicitly opt-in in contrary to the folders stated in the description.

```js
defineNuxtConfig({
  extends: [
    './base',
    {
      eslint: true, // reads eslint.config extensions
      prettier: true, // reads prettier extensions
     }
  ]
})
```

or maybe explicit `files` array to avoid the need for handling resolving config extensions
```js
defineNuxtConfig({
  extends: [
    './base',
    {
      files: [ '.editorconfig', '.eslintrc.cjs', '.prettierrc' ]
     }
  ]
})
```

**Comment by localusercamp** [comment]
> My use case is I want to have a base repo that has core UI components in addition to eslint, prettier configs.\nI want to use those UI components in addition to eslint.config and .prettierrc in app-1, app-2 repos for example to ensure code consistency (I think this is much like a mono repo?).

@MuhammadM1998 What about making eslint/prettier monorepo packages (as a workaround). You still have to add config files manually, but they will consist of requiring your package/plugin only.
1. Providing code style conventions (eslint, prettier, etc) through UI Layer - is a bit strange because this is not a responsibility of the UI Layer.
2. Every environment (layer) must contain those config files "physically" because of IDE support (that means that you cannot do it with JS, those files just must be there anyway).

**Comment by MuhammadM1998** [comment]
@localusercamp Thanks for clarifying you gave me ideas. ``I'm exploring the monorepo approach. What made me ask is layers seems simpler to achieve my goal than the monorepo. Layers is not for UI only it can contain shared logic and utils too so i think it's okay For my use case.  As for the 'files needs to be physically in root' bit its done with 'tailwind.config' in nuxt tailwind module for example, idk if it's doable in a layer though

**Comment by localusercamp** [comment]
> Layers is not for UI only it can contain shared logic and utils too so i think it's okay For my use case

I'm not saying that they could contain **only** UI components. What I want to say here is that in **your case** you have a UI layer as you call it, so if your layer consists of UI components it's a bit strange that it also contains prettier config that must be **exposed** because this breaks the SRP rule. I think in that case your UI layer must depend on your prettier package.

> As for the 'files needs to be physically in root' bit its done with 'tailwind.config' in nuxt tailwind module for example.

You are right, but tailwindcss module runs on the build step. Prettier/eslint runs in IDE.

**Comment by MuhammadM1998** [comment]
@localusercamp Aah I understand your viewpoint now. Thanks a lot this is really helpful 🙏

**Comment by CptJJ** [comment]
Im still having issues with the /public not being extended when deploying on vercel. works fine in dev! Some of the linked issues seems to be closed but seeing the problem in nuxt 3.8.0

**Comment by blowsie** [comment]
> Im still having issues with the /public not being extended when deploying on vercel. works fine in dev! Some of the linked issues seems to be closed but seeing the problem in nuxt 3.8.0

As far as I can see that's not in the scope 
https://nuxt.com/docs/guide/going-further/layers

> Additionally, certain other files in the layer directory will be auto-scanned and used by Nuxt for the project extending this layer.
[components/*](https://nuxt.com/docs/guide/directory-structure/components) - Extend the default components
[composables/*](https://nuxt.com/docs/guide/directory-structure/composables) - Extend the default composables
[pages/*](https://nuxt.com/docs/guide/directory-structure/pages) - Extend the default pages
[server/*](https://nuxt.com/docs/guide/directory-structure/server) - Extend the default server endpoints & middleware
[utils/*](https://nuxt.com/docs/guide/directory-structure/utils) - Extend the default utils
[nuxt.config.ts](https://nuxt.com/docs/guide/directory-structure/nuxt-config)- Extend the default nuxt config
[app.config.ts](https://nuxt.com/docs/guide/directory-structure/app-config) - Extend the default app config

Regarding scope, does it make sense to export `stores/*` for those using pinia?

**Comment by AndriiGera** [comment]
While extending nuxt apps is a really nice feature and active improvement looks promising, still one use case seems to be missing - the ability to combine mutiple nuxt layers into a multiapp. 
For example, I have an app Main and N number of custom themes that only override css and optionally few components/layouts of Main.
I would like to have a single Nuxt app that will extend Main and all Theme layers:
![Nuxt Layers drawio](https://github.com/nuxt/nuxt/assets/93651428/14cc143a-4272-4bfc-94dd-3cd2e2189ca9)

and will somehow output every theme to a specific subfolder in `.output`.

Currently, with the help of layers I can extend Main app with N amount of Themes and have them all as separate Nuxt instances  which is not very scalable in terms of build process and hosting as the amount of themes can increase.

**Comment by pi0** [comment]
@AndriiGera Multi-app via multi-build is of course doable. Each app can extend from shared base theme(s). Of course yo should be then responsible to handle cross-app routing, state preservation and there are duplicate date in bundle (vue is bundled in each app). We can configure nitro to also scope `.outout` dir for each.

It is not original purpose of multi-app via layers (we prefer single build with everything including style and layouts fully scoped) but if you like to followup with documenting it, feel free to make a new discussion/issue and ping me 👍🏼

**Comment by akasection** [comment]
So I have a question regarding `definePageMeta` and layers. In my company, we are trying to adopt layers as installable features so different departments can create whole set of feature and then with some configuration via bundled modules, they can pass and alter the feature to suit the consumers.

But now we have stumbled upon `definePageMeta` and the correct way to override the layer pages meta from the consumer app. Let's say `feature-a` (published as npm package) is installed into main `product-app`, but we want to apply middleware into all `feature-a` pages and also alter the page URL so it resides to `feature-alpha` instead of `feature-a` in the `product-app`. Now, since `definePageMeta` is a compiler macro, we can't just make "props" for the page.

In the end, we have three kinds of approaches that are used across our company. I have summarized the way to do it in this CodeSandbox: https://codesandbox.io/p/sandbox/nuxt-layer-pagemeta-example-7wws6l

In summary, the methods are:
1. Method A: Use `pages:extend` hook to alter the `page` property
2. Method B: Use `AppConfig` and anonymous middleware
3. Method C: Re-template the vue page file, "insert" new `definePageMeta` as string

The question is, **what is the proper way** to do this kind of thing? What's the recommended approach to alter layer parts from consumer app without making global object/function that intercept/checking things?

**Comment by SunSi12138** [comment]
when I write type and interface in types directory, there's no need to export or import, I can use it in everywhere , just like magic. And I dont know why.

But when one project can't use types defined in it's layer, I wonder if there's a way to fix it?

I really like this clean way to define types

**Comment by SunSi12138** [comment]
when  I install module and package in base layer,  the second layer won't run and throws this exception :

` yarn run dev
yarn run v1.22.19
$ nuxt dev
Nuxt 3.8.2 with Nitro 2.8.1                                                14:47:00
                                                                           14:47:00
  ➜ Local:    https://localhost:3000/
  ➜ Network:  use --host to expose


[14:47:01]  ERROR  Error while requiring module @pinia/nuxt: Error: Cannot find module 'F:/Project/App/@pinia/nuxt'
Require stack:
- F:\Project\App\index.js


[14:47:01]  ERROR  Cannot start nuxt:  Cannot find module 'F:/Project/App/@pinia/nuxt'
Require stack:
- F:\Project\App\index.js

  Require stack:
  - index.js
  at Module._resolveFilename (node:internal/modules/cjs/loader:1075:15)
  at Function.resolve (node:internal/modules/cjs/helpers:116:19)
  at Function._resolve [as resolve] (node_modules\jiti\dist\jiti.js:1:251148)        
  at resolveModule (/F:/Project/App/node_modules/@nuxt/kit/dist/index.mjs:2239:29)   
  at requireModule (/F:/Project/App/node_modules/@nuxt/kit/dist/index.mjs:2244:24)   
  at loadNuxtModuleInstance (/F:/Project/App/node_modules/@nuxt/kit/dist/index.mjs:2472:90)
  at async installModule (/F:/Project/App/node_modules/@nuxt/kit/dist/index.mjs:2438:47)
  at async initNuxt (/F:/Project/App/node_modules/nuxt/dist/index.mjs:3702:7)        
  at async NuxtDevServer._load (/F:/Project/App/node_modules/nuxi/dist/chunks/dev2.mjs:253:5)
  at async NuxtDevServer.load (/F:/Project/App/node_modules/nuxi/dist/chunks/dev2.mjs:185:7)`
  
  and I have to reinstall it on layer two
  
  any suggestions?

**Comment by ojvribeiro** [comment]
@SunSi12138 dependencies are still not auto-installed from layer. You need to install all your layer dependencies manually on your project.

The Nuxt team is working on an auto-install for layers deps and it's being tracked here: https://github.com/unjs/c12/issues/51

**Comment by Arthur-Chevron** [comment]
HI, I have three layers, one is for UI, one is for LOGIC and the last is for my main app.
I am configuring all my types for my components in the repo UI but i need to have access in my main app too. 
What is the best way to deal with it.

**Comment by jofmi** [comment]
Would it be possible to add support for `error.vue` to this list?

**Comment by hamedfaryabi** [comment]
How can I import a component?

I have a component called ElementsIcon in my layer and I want to use this component in the h() function in my project. Like the following code:
``` typescript
const icon = h(ElementsIcon, {
    name: 'dismiss',
    size: '20px',
    'v-tooltip': { content: __('cancel order'), placement: 'bottom', distance: 12 },
	onClick: () => {
	onCancelOrder(row.id);
    }
});
```
But I don't know how to import this component

**Comment by enkot** [comment]
@hamedfaryabi You can import it from `#components`:
```ts
import { ElementsIcon } from '#components'
```
But better to ask questions here https://github.com/nuxt/nuxt/discussions/categories/questions or in the [Discord](https://discord.com/invite/nuxt)

**Comment by productdevbook** [comment]
`stores` or `newFolder` How can I add a customized folder ?

**Comment by ojvribeiro** [comment]
Now that https://github.com/unjs/c12/issues/51 is closed, I'm so hyped for it to land on Nuxt. ♥

**Comment by pi0** [comment]
> Now that https://github.com/unjs/c12/issues/51 is closed, I'm so hyped for it to land on Nuxt. ♥

You can already do this!

```js
{
  extends: [ 
    ["foobar", { install: true }]
  ]
}
```

This is opt-in behavior until a while we can test all edge cases and optimize installation speeds.

**Comment by KeJunMao** [comment]
How do I overwrite layer plugins in my app?

Currently, they are both valid, and even if the plugin name is specified, it cannot be overwritten
```
map-app
  plugins
    setup.base.ts

base-layer
  plugins
    setup.base.ts
```

**Comment by baixiaoyu2997** [comment]
I want my base layer to run independently and be testable，so i need add `"extends": ["./.nuxt/tsconfig.json"]` to `tsconfig.json` in base layer, but this causes a problem, I need to run prepare everywhere,  hope this process can be simplified
![image](https://github.com/nuxt/nuxt/assets/9149999/4a588dd0-4eaa-4322-b998-8575b09bc36f)
![image](https://github.com/nuxt/nuxt/assets/9149999/3f6fe91d-2e8e-45c8-aa12-cec42b9d6b7d)

**Comment by jofmi** [comment]
I have a question regarding layers and typescript.

If I put the following in `layer1/index.d.ts`:

```ts
declare global {
  interface MyType {
    attr1: string
  }
}

export {};
```

And the following in `layer2/index.d.ts`:

```ts
declare global {
  interface MyType {
    attr2: string
  }
}

export {};
```

If `layer2` extends `layer1`, typescript in `layer2` will then interpret the following combined type:

```ts
interface MyType {
    attr1: string
    attr2: string
}
```

I find this incredibly useful, but wanted to ask whether this is intended behavior or whether I am using a weird hack - and whether there is a danger that this will stop working in the future?

**Comment by kjfranke** [comment]
@pi0 / @danielroe I'm not sure what the progress is on "Named layer aliases", I created a Nuxt module that creates an alias per layer and made ~,@, etc aliases fall back from project down to each layer. I'm not sure if every case is covered but it works for our use cases. If you are interested, I'm happy to share. But I do belief this feature should exist native in Nuxt instead of a separate module.

**Comment by odranoelBR** [comment]
Hi @kjfranke , can you share it ?

**Comment by adamdehaven** [comment]
Is there any plan to properly resolve types in general, especially generated when using layers in a pnpm monorepo? 

If this has been solved (it's not on the list here) I can't seem to find it?

**Comment by mts-pac** [comment]
> css files from base layer are added to the html head after the project css files.
> 
> please see: [stackblitz.com](https://stackblitz.com/edit/nuxt-starter-mvptjc?file=nuxt.config.ts)
> 
> nuxt.config.ts:
> 
> ```
> export default defineNuxtConfig({
>   devtools: { enabled: false },
>   css: ['~/assets/css/project.css'],
>   extends: ['./layers/base'],
> });
> ```
> 
> base layer nuxt.config:
> 
> ```
> import { fileURLToPath } from 'url';
> import { dirname, join } from 'path';
> 
> const currentDir = dirname(fileURLToPath(import.meta.url));
> 
> export default defineNuxtConfig({
>   css: [join(currentDir, './assets/css/base.css')],
> });
> ```
> 
> in the website source:
> 
> ```
> <head>
> <link rel="stylesheet" href="/_nuxt/assets/css/project.css">
> <link rel="stylesheet" href="/_nuxt/layers/base/assets/css/base.css">
> </head>
> ```
> 
> project.css should clearly be after base.css file(s) so that project can overwrite styles?

Any solution?

**Comment by Brads3290** [comment]
Layers do not seem to install dev dependencies correctly.

I created #26980 for this, and have now found a few other cases where people are saying that #19124 is still present ([here](https://github.com/nuxt/nuxt/issues/26479), [here](https://github.com/nuxt/nuxt/issues/19124#issuecomment-1991685386), and [here](https://github.com/nuxt/nuxt/issues/19124#issuecomment-2001921887)), and all of those cases are _specifically_ dev dependencies. Runtime dependencies seem to work fine.

Not sure what the repercussions would be of indiscriminately installing all dev dependencies from layers, but it does seem important that Nuxt modules at the very least are installed (`@nuxt/tailwindcss`, `@nuxt/ui`, for example).

**Comment by adamdehaven** [comment]
> Layers do not seem to install dev dependencies correctly.

I commented on the issue you created. This seems like expected behavior: https://github.com/nuxt/nuxt/issues/26980#issuecomment-2081661835

**Comment by Brads3290** [comment]
Yeah, I think you're right that at the moment it is expected behaviour - I shouldn't have called it a bug.

However, I think it needs discussion, because right now it seems installing modules in a layer will make that layer unusable. I suspect this happens because:
1. The module installs as a dev dependency (this happens automatically when added via `npx nuxi@latest module add <module>`)
2. The main app must look at the layer's `"modules": [...]` key in nuxt.config.ts, and expect to find the specified modules installed
3. Because the module was installed as a dev dependency, and dev dependencies aren't inherited, it falls over (can no longer `npm install` or run the Nuxt server from the main app)

**Comment by yuriystasiv** [comment]
> > css files from base layer are added to the html head after the project css files.
> > please see: [stackblitz.com](https://stackblitz.com/edit/nuxt-starter-mvptjc?file=nuxt.config.ts)
> > nuxt.config.ts:
> > ```
> > export default defineNuxtConfig({
> >   devtools: { enabled: false },
> >   css: ['~/assets/css/project.css'],
> >   extends: ['./layers/base'],
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > base layer nuxt.config:
> > ```
> > import { fileURLToPath } from 'url';
> > import { dirname, join } from 'path';
> > 
> > const currentDir = dirname(fileURLToPath(import.meta.url));
> > 
> > export default defineNuxtConfig({
> >   css: [join(currentDir, './assets/css/base.css')],
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in the website source:
> > ```
> > <head>
> > <link rel="stylesheet" href="/_nuxt/assets/css/project.css">
> > <link rel="stylesheet" href="/_nuxt/layers/base/assets/css/base.css">
> > </head>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > project.css should clearly be after base.css file(s) so that project can overwrite styles?
> 
> Any solution?

I was able to have it done, by importing css for base layer in `nuxt.config.ts:`

```
import { createResolver } from '@nuxt/kit';

const { resolve } = createResolver(import.meta.url);

export default defineNuxtConfig({
  alias: { '#base': resolve('./') },
  css: ['#base/assets/css/base.css'],
});
```

and next instead of importing css in `nuxt.config.ts`, import inside `app.vue` script or style tag:

```
<script>
// this works
import '~/assets/css/project.css';
</script>

<!-- this should also work -->
<!-- <style>
@import '~/assets/css/project.css';
</style> -->
```

Please see: https://stackblitz.com/edit/nuxt-starter-tcptzx?file=app.vue

**Comment by yuriystasiv** [comment]
The same for scss: https://stackblitz.com/edit/nuxt-starter-orbtho?file=app.vue

**Comment by haelbichalex** [comment]
For anyone searching the module @kjfranke mentioned, it's located here: https://github.com/kjfranke/nuxt3-layer-alias

For me it looks like a good option if you want to share and/or overwrite f.e. `assets` between layers.

**Comment by c0nd3v** [comment]
Waiting for multi app support

**Comment by igorjacauna** [comment]
Hey! Layers are amazing! Is there any possibility of using them with the content directory from @nuxt/content? Or somehow specifying which directories from the layers should be merged? Or this is a bad idea?

Thank you!

**Comment by mavyfaby** [comment]
any updates on these?

**Comment by DavidSabine** [comment]
Suggestion:

One or both of these pages could include a description of how to make a nuxt layer from scratch.

* https://nuxt.com/docs/getting-started/layers
* https://nuxt.com/docs/guide/going-further/layers

Every example I can find is either: "just put a nuxt.config.js into a subfolder in the "app" directory and you have a new layer" or "init a new project using the `layer` template"

But given we're between v3 and v4, I feel the need to really understand how a 'layer' is different from an 'app'. Is it just a basic nuxt app minus the app.vue file? plus a .playground folder? then `nuxi prepare`?
(No response needed to those questions... those are rhetorical.)

**Comment by rasliche** [comment]
> Suggestion:
> 
> One or both of these pages could include a description of how to make a nuxt layer from scratch.
> 
> * https://nuxt.com/docs/getting-started/layers
> * https://nuxt.com/docs/guide/going-further/layers
> 
> Every example I can find is either: "just put a nuxt.config.js into a subfolder in the "app" directory and you have a new layer" or "init a new project using the `layer` template"
> 
> But given we're between v3 and v4, I feel the need to really understand how a 'layer' is different from an 'app'. Is it just a basic nuxt app minus the app.vue file? plus a .playground folder? then `nuxi prepare`? (No response needed to those questions... those are rhetorical.)

Hey @DavidSabine, I think you've got it and need to try to not overthink it. My understanding is Layers truly are just a Nuxt app that you can merge into another app and then overwrite anything you need to. Also not sure what you mean about between v3 and v4, but my take there is that the Nuxt core devs have said don't expect it to be a big change like from 2 -> 3. I think they said it's mainly to make things easier and faster, but not changing how things work in general.

I agree though that sometimes the docs feel duplicative (like how you linked the getting started and going further pages) and also kind of high level. There are some awesome Nuxt devs that share knowledge for free (Alex Lichter and Daniel Roe jump out as being especially prolific and accessible, but there are definitely others like Michael Thiessen's course).  But sometimes it feels like the basics get outsourced too much.

Not to derail your question, but I'd add that I still go back and forth between whether I should write a module or a layer. I feel like that distinction is particularly unclear, other than "You probably want a Layer, Modules are complicated".

**Comment by Pentadome** [comment]
is there a way to encapsulate the css of a layer? So it is not used in other layers?

**Comment by Dominic-Marcelino** [comment]
> is there a way to encapsulate the css of a layer? So it is not used in other layers?

You could add a base-class to a top-level element and then scope the css to it.

**Comment by Anoesj** [comment]
Not sure if I'm running into a known issue here, but I have an application fully built with layers (almost no top-level files, everything in a layer). I have turned on `typedPages: true`, and I can see the `RouteNamedMap` interface is generated correctly in `.nuxt/types/typed-router.d.ts`, but nor `navigateTo`, nor `$router.push`/`$router.replace` in templates, nor `useRouter().push`/`useRouter().replace` are using the typed routes. It basically just accepts any route name. When I enter a valid route name, it still doesn't recognize it or its route params.

If it's not a known issue, I'll happily create a reproduction 🙂.

**Comment by drowhannn** [comment]
I couldn't find any way to ignore some files while extending a layer. A way to ignore files or selectively extend needed files would be nice.

**Comment by mgd216** [comment]
Is there planed support for allowing the 'shared' folder to be extended through layers?

For example, I have a nuxt-base layer with shared/utils/fmt-date.ts

```
import { format } from 'date-fns';

/**
 * Get basic formatted date 01/31/2021
 *
 * @param {number} timestamp
 *
 * @returns {string} 01/31/2021
 */
export default function (timestamp?: number | null): string {
  if (!timestamp) return '';
  return format(timestamp, 'MM/dd/yyyy');
}
```
The fmtDate() function is 'auto imported' in the nuxt-base app (which is awesome BTW).
When I 'extend' the nuxt-base layer in my nuxt 'home' app, the fmtDate() function is not autoimported.

The Nuxt Layers [Docs](https://nuxt.com/docs/getting-started/layers) show that it is currently only the components, composables and utils, diretory.  It would be great if the extended layer also included 'shared/types' and 'shared/utils'.  Thanks!

**Comment by MaciejJezierskiFiskars** [comment]
Is there any plan to support modules folder coming from layer?

**Comment by sandros94** [comment]
> Is there any plan to support modules folder coming from layer?

it is already supported, both on local and remote layers. Do you have any particular issue?

**Comment by MaciejJezierskiFiskars** [comment]
@sandros94 I have an issue after switching from pnpm to bun (due to problems with monorepo and c12 autoinstall). After switch local modules are not recognised. I will try to reproduce it and will create separate issue.

**Comment by sandros94** [comment]
> [@sandros94](https://github.com/sandros94) I have an issue after switching from pnpm to bun (due to problems with monorepo and c12 autoinstall). After switch local modules are not recognised. I will try to reproduce it and will create separate issue.

Feel free to tag me in the new issue

**Comment by mrleblanc101** [comment]
+1 for `shared/types` and `shared/utils` in Layers, it doesn't seem to work.
`#shared` will resolved to `./.playground/shared` and not `./shared`.
`@@/shared` will resolve to `rootDir` and `@` to `srcDir` of the `.playground`, not the actual layer

**Comment by DavidSabine** [comment]
Hi @rasliche,

(Just seeing your comment now.)

For clarification regarding the comment "between v3 and v4": I meant to draw attention to the fact that the folder structure has changed in v4 - so how to override things inherited from a layer can get tricky. And I've seen comments from others that some things break (or misbehave) with a v4 Nuxt app inherits a v3 layer, and vice versa.

**Comment by serhii-chernenko** [comment]
As I got it, the `shared` directory is not about sharing types or utils between layers. It just allows sharing the same util between Vue components and Nitro server functions. The info is mentioned [here](https://nuxt.com/docs/guide/directory-structure/shared).

But I'm struggling with the same issue. I have a few layers and end application.
For example:
```
root/
- apps/
    - app
- layers/
    - ui
    - core 
```

So, I use the `dev:prepare` command to build everything to `.playground/.nuxt` per each layer. 
And then `dev` for the end application in `apps/app`.
For example, I use some components, types and composables in the `ui` layer. They are resolved for the `ui` layer and for the run end app. But when I edit files in `layers/core` the imports are not resolved automatically in PHPStorm and VS Code.

Nuxt works as expected without any errors, but the IDE and code editor don't understand what to import because the same imports with different paths are existing in different `.nuxt` apps:
```
apps/app/.nuxt/imports.d.ts
layers/ui/.playground/.nuxt/imports.d.ts
```

Real code examples:
![Image](https://github.com/user-attachments/assets/3d10ae21-d911-4507-9f35-2ac4fad67c5b)
![Image](https://github.com/user-attachments/assets/bbe8daf8-516c-4263-882a-385574f7b6ec)

The same for types, I have to specify package (in monorepo) form where to get the type because it's not resolved.

I started the app in VS Code, but then it doesn't work well with shared ESLint configs. I switched to PHPStorm, and it works much better with ESLint shared config in monorepo. But some situations like this make me sad because it's difficult to set up, no examples, no simple solution. It takes many time readings issues, discussions, researching any examples, asking AI assistants, etc.

The layers' functionality is a really powerful tool. But please make it easy for Vue and Nuxt ecosystem in general. Currently, it beats developers more than provide advantages.

The documentation has to provide detailed instructions:
- How to set up shared ESLint between all layers for the most popular editors: PHPStorm (or WebStorm) and VS Code.
- How to prepare a monorepo and application with many layers.
- How to share components, utils, composables and types between all layers and make auto-imports working correctly for the mentioned code editors.
- How to correctly include (extend) layers outside the application directory (especially for monorepo).
- Any tools `$` provided from plugins have to be supported and resolved correctly for the code editors.
- It'd be nice to have at least simple examples of how to work with databases included as a separate layers and be shared between others. I, personally, appreciate any cases with a couple of DBs.

The whole Nuxt documentation is perfect, but the layers area feels like a gap there.

**Comment by dmartel** [comment]
Hello !

I'm trying out Layers, and I'm encountering a routing challenge. My setup is as follows:

```javascript
// MyApp configuration
extends: [
  '../app2',
  '../app1',
  '../ui',
];
```

Both **app1** and **app2** are applications with their own routing:
app1 has routes for "home" and "events"
app2 has routes for "home" and "items"
These work perfectly fine when run independently. 

However, in my main application **myApp**, I'm trying to achieve the following routing structure:
```text
http://localhost:3000/app1/events
http://localhost:3000/app2/items
```
Currently, when I click on the "events" link from my **app1** component in **myApp**, I'm redirected to http://localhost:3000/events (instead of http://localhost:3000/app1/events).

_Questions:_
Is it possible to implement this kind of nested routing structure with Layers ?
How can I set up the router to maintain the /app1 and /app2 prefixes for their respective routes?
This approach would also help avoid route conflicts between app1 and app2 (e.g., both have a "home" route). Any suggestions on best practices for handling this?
I'd appreciate any insights or solutions to implement this routing architecture effectively. 

Thanks in advance for your help!

**Comment by mrleblanc101** [comment]
@dmartel Layers get merged, so if you do `/layer1/pages/events.vue` and `/layer2/pages/events.vue`, your final build will have one single root event page `/app/pages/events.vue` and which one it is will depend on the order of your extends.

The route name `events` without prefix can reference any of `app/pages/events.vue`, `layer1/pages/events.vue` or `layer2/pages/events.vue`.

To do what you want, you'll need to create `/layer1/pages/layer1/events.vue` and `/layer2/pages/layer2/events.vue` and your NuxtLink will need to use the proper route name `layer1-events` and `layer2-events`.

**Comment by dmartel** [comment]
Thanks for the explanations @mrleblanc101 

I was trying to find a way to have my layer1 routes prefixed with the layer name but today it doesn't exist.
But  we have to create a subfolder in "pages" to force routing like this 👍

**Comment by mrleblanc101** [comment]
@dmartel There might be a way to do it, but I wouldn't bother as it goes against how layer are supposed to work (merging)

**Comment by gretchelin** [comment]
HI, is there a plan to have `assets` folder "shared" with other layers?

I had expected Nuxt to merged layers' `assets` folder to root, the way `public` folder is being handled, but it doesn't seem so. I got issue - basically asset file is not found - when running project from root and root `assets` folder does not have the assets in it.

Or maybe there's a recommended way to have assests folder automatically imported/referred correctly in other layer?

**Comment by kjfranke** [comment]
@gretchelin I'm not sure what the current official Layers/Extend status is. There were plans for named layers which might solve this. In the meanwhile here is a basic module I wrote which change the behavior of aliases by searching for the target file in every layer: So ~/assets/images/logo.svg will check the main project first and drills down each layer.

Besides that, it creates a new alias for each layer (based on the folder name) so you can target specific layers.

It's far form perfect and probably not the most efficient way of doing this. In the readme you find details on how it works so you can modify it even further to suit your project.

I hope it helps you out.

https://github.com/kjfranke/nuxt3-layer-alias/
